### PR TITLE
Fix RegExp.test, closes #74

### LINKED ISSUE #74: regex test is not pure
`regex.test` on global or sticky regular expressions will mutate the `lastIndex`, therefore using `test` is not a pure function.

For example this is what sanctuary [is doing](https://github.com/sanctuary-js/sanctuary/blob/71f138cbe02382f9b3f46bdd387919cfa8bc2f39/index.js#L4404-L4409) and [this](https://github.com/sanctuary-js/sanctuary/blob/71f138cbe02382f9b3f46bdd387919cfa8bc2f39/index.js#L4465-L4469)

I suspect we should do something similar here too for example this is what I have:
```ts
/**
 * Remembers `lastIndex` of the provided `regex` and resets it back after running the thunk
 */
export function withRegex<T>(regex: RegExp, thunk: () => T): T {
  const lastIndex = regex.lastIndex;
  const result = thunk();
  regex.lastIndex = lastIndex;
  return result;
}
/**
 * A pure function version of regex.test(s). i.e. this will not change the `lastIndex` prop of the regex.
 * see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/test#using_test_on_a_regex_with_the_global_flag
 */
export function regexTest(regex: RegExp, s: string) {
  return withRegex(regex, function () {
    return regex.test(s);
  });
}
```

### PR DESCRIPTION
**Pull request by DenisFrezzato**

### CODE REVIEW

File: src/RegExp.ts [hunk 1]
```diff
@@ -39,7 +39,14 @@ export const match: (r: RegExp) => (s: string) => O.Option<RegExpMatchArray> = (
  *
  * @since 0.1.8
  */
-export const test: (r: RegExp) => Predicate<string> = (r) => (s) => r.test(s)
+export const test: (r: RegExp) => Predicate<string> = (r) => {
+  const lastIndex = r.lastIndex
```
Comment (safareli) [inline src/RegExp.ts:43]: This should be done immediately before the `r.test` call
```
const regex = /,/g
const tester = test(regex)
regex.lastIndex = 2
tester(",asdasd")
console.log(regex.lastIndex) // 0 - should be 2
```

**Comment by safareli** [comment]
👍

**Review by gcanti — APPROVED** [review]

**Comment by gcanti** [comment]
Thanks @DenisFrezzato @safareli, patch released.

**Pull request merged**